developer_guides/rimage/ipc4_extended_module_config.rst

```diff
@@ -0,0 +1,247 @@
+.. _ipc4_extended_module_config:
+
+Ipc4_extended_module_config
+###########################
+
+The ipc4 extended module config is a place to store build-time known module metadata
+such as the module name, uuid or capabilities. This config is part of the output
+binary is located in extended manifest structure in the binary file and it is used
+by driver when setting up ipc message to dsp.
+
+Rimage builds extended module config based on module setting in config/xxx.toml.
+The following setting is for gain module.
+
+.. code-block:: bash
+
+        [[module.entry]]
+        name = "GAIN"
+        uuid = "61BCA9A8-18D0-4A18-8E7B-2639219804B7"
+        affinity_mask = "0x1"
+        instance_count = "40"
+        domain_types = "0"
+        load_type = "0"
+        module_type = "5"
+        auto_start = "0"
+        sched_caps = [1, 0x00008000]
+
+        # pin = [dir, type, sample rate, size, container, channel-cfg]
+        pin = [0, 0, 0xfeef, 0xf, 0xf, 0x45ff,
+                        1, 0, 0xfeef, 0xf, 0xf, 0x1ff]
+
+        # mod_cfg [PAR_0 PAR_1 PAR_2 PAR_3 IS_BYTES CPS IBS OBS MOD_FLAGS CPC OBLS]
+        mod_cfg = [0, 0, 0, 0, 416, 914000, 48, 64, 0, 0, 0,
+                                1, 0, 0, 0, 416, 1321600, 192, 256, 0, 0, 0,
+                                2, 0, 0, 0, 416, 1786000, 192, 256, 0, 0, 0,
+                                3, 0, 0, 0, 416, 2333000, 48, 64, 0, 0, 0,
+                                4, 0, 0, 0, 416, 2910000, 192, 256, 0, 0, 0,
+                                5, 0, 0, 0, 416, 3441000, 192, 256, 0, 0, 0,
+                                6, 0, 0, 0, 416, 4265000, 192, 256, 0, 0, 0]
+
+Module entry definition
+***********************
+| **name**: name of the module
+| **uuid**: uuid of the module
+| **affinity_mask**: usable dsp core mask, bit N is for core N, 0x3 means core 0 and 1 can be used
+| **instance_count**: max module instance count can be active in fw
+| **domain_types**: two schedule domains: domain ll (0), and dp (1)
+| **load_type**: build-in module (BIT(0)), loadable (BIT(1)) or both BIT(0) | BIT(1)
+| **module_type**: defined in the following enum module_type
+
+.. code-block:: bash
+
+  enum module_type {
+        ebasefw = 0,
+        emixin,
+        emixout,
+        ecopier,
+        epeakvol,
+        eupdwmix,
+        emux,
+        esrc,
+        ewov,
+        efx,
+        eaec,
+        ekpb,
+        emicselect,
+        efxf,   /*i.e.SmartAmp */
+        eaudclass,
+        efakecopier,
+        eiodriver,
+        ewhm,
+        egdbstub,
+        esensing,
+        emax,
+        einvalid = emax
+   };
```

Comment:
@mwasko @mmaka1 Can we add something about the semantics here? What should people adding new modules know about setting the "module_type"?

Reply:
@kv2019i module_type enum was defined for the Host SW driver purpose. The idea was to have option to identify components of specific functionality like for e.g. wov, mux or any other listed above. That capability allow for more topology flexibility at Host SW level. For example, instead of hardcoding in topology that you need Module_1 with specific GUID you can point that you need Module_1 with specific functionality type. That allow to swap components of the same functionality without need to change the topology itself / modifying GUIDs.

The bottom line is that new modules module_type should be matched with one of the available on list or if necessary the new custom type should be added. I would also recommend to change the enum name to something like module_fuctionality_type.

Reply:
Ack @mwasko . I think the "bottom line" part should be added directly in the docs to help module developers choose correctly.